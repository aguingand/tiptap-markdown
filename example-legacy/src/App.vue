<script setup>
    import Editor from "./components/Editor.vue";
</script>

<template>
    <div class="container">
        <Editor />
    </div>
</template>

<script>
    import 'bootstrap/dist/css/bootstrap.css';
</script>

<style lang="scss" src="./app.scss">
</style>
