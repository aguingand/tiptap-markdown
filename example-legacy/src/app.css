

.hljs-comment,
.hljs-quote {
    color: #616161;
}

.hljs-variable,
.hljs-template-variable,
.hljs-attribute,
.hljs-tag,
.hljs-name,
.hljs-regexp,
.hljs-link,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class {
    color: #F98181;
}

.hljs-number,
.hljs-meta,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params {
    color: #FBBC88;
}

.hljs-string,
.hljs-symbol,
.hljs-bullet {
    color: #B9F18D;
}

.hljs-title,
.hljs-section {
    color: #FAF594;
}

.hljs-keyword,
.hljs-selector-tag {
    color: #70CFF8;
}

.hljs-emphasis {
    font-style: italic;
}

.hljs-strong {
    font-weight: 700;
}
