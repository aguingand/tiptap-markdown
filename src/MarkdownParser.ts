import { Content, Editor, getExtensionField, MarkConfig, NodeConfig } from "@tiptap/core";
import { unified } from "unified";
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import rehypeMinifyWhitespace from 'rehype-minify-whitespace';
import { Root } from 'mdast';
import rehypeDomStringify from "rehype-dom-stringify";

export class MarkdownParser {
    editor: Editor;

    constructor(editor: Editor) {
        this.editor = editor;
    }

    parse(content: Content): Content {
        if(typeof content === 'string') {
            const fromMarkdown = unified()
                .use(remarkParse);
            const toHTML = unified()
                .use(remarkRehype)
                .use(rehypeMinifyWhitespace)
                .use(rehypeDomStringify)

            this.editor.extensionManager.extensions.forEach(extension => {
                const parseMarkdown = getExtensionField<NodeConfig['parseMarkdown'] | MarkConfig['parseMarkdown']>(
                    extension,
                    'parseMarkdown',
                    {
                        name: extension.name,
                        options: extension.options,
                        editor: this.editor,
                    }
                );
                parseMarkdown?.({
                    fromMarkdown,
                    toHTML,
                });
            });
            const mdast = fromMarkdown.runSync(fromMarkdown.parse(content), content) as Root;
            const hast = toHTML.runSync(mdast, content);
            if(hast.children.length === 0) {
                // earle return if empty because of https://github.com/rehypejs/rehype-dom/pull/26
                return '';
            }
            const html = toHTML.stringify(hast, content);
            return html;
        }

        return content;
    }
}
