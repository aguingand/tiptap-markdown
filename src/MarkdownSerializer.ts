import remarkStringify from 'remark-stringify';
import { Editor, getExtensionField, getHTMLFromFragment, MarkConfig, NodeConfig } from "@tiptap/core";
import { Node as ProseMirrorNode, Fragment } from "@tiptap/pm/model";
import { unified } from "unified";
import rehypeRemark from "rehype-remark";
import rehypeDomParse from "rehype-dom-parse";
import { Root } from "hast";

export class MarkdownSerializer {
    editor: Editor;

    constructor(editor: Editor) {
        this.editor = editor;
    }

    serialize(content: ProseMirrorNode | Fragment) {
        return this.serializeFromHTML(
            getHTMLFromFragment(Fragment.from(content), this.editor.schema)
        );
    }

    serializeFromHTML(content: string): string {
        const fromHTML = unified()
            .use(rehypeDomParse, { fragment: true });
        const toMarkdown = unified()
            .use(rehypeRemark)
            .use(remarkStringify);

        this.editor.extensionManager.extensions.forEach(extension => {
            const renderMarkdown = getExtensionField<NodeConfig['renderMarkdown'] | MarkConfig['renderMarkdown']>(
                extension,
                'renderMarkdown',
                {
                    name: extension.name,
                    options: extension.options,
                    editor: this.editor,
                }
            );
            renderMarkdown?.({
                fromHTML,
                toMarkdown,
            });
        });

        const hast = fromHTML.runSync(fromHTML.parse(content), content) as Root;
        const mdast = toMarkdown.runSync(hast, content);
        const md = toMarkdown.stringify(mdast, content);
        return md;
    }
}
