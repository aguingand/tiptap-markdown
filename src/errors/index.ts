

const functionReplacer = (_: string, value: unknown) => (typeof value === 'function' ? '[Function]' : value)

const stringify = (x: unknown): string => JSON.stringify(x, functionReplacer)

export function createNodeInParserFail(...args: unknown[]) {
    // todo handle args
    return new Error('Create prosemirror node from remark failed in parser');
}

export function parserMatchError(node: unknown) {
    return new Error(`Cannot match target parser for node: ${stringify(node)}.`);
}

export function serializerMatchError(node: unknown) {
    return new Error(`Cannot match target serializer for node: ${stringify(node)}.`);
}

export function stackOverFlow() {
    return new Error('Stack over flow, cannot pop on an empty stack.');
}
