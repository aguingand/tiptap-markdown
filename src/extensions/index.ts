import Blockquote from "./extension-blockquote/blockquote";
import BulletList from "./extension-bullet-list/bullet-list";
import CodeBlock from "./nodes/code-block";
import Document from "./extension-document/document";
import HardBreak from "./nodes/hard-break";
import Heading from "./nodes/heading";
import HorizontalRule from "./nodes/horizontal-rule";
import Image from "./nodes/image";
import ListItem from "./nodes/list-item";
import OrderedList from "./nodes/ordered-list";
import Paragraph from "./extension-paragraph/paragraph";
import Table from "./nodes/table";
import TaskItem from "./nodes/task-item";
import TaskList from "./nodes/task-list";
import Text from "./extension-text/text";

import Bold from "./extension-bold/bold";
import Code from "./extension-code/code";
import Italic from "./extension-italic/italic";
import Link from "./extension-link/link";
import Strike from "./extension-strike/strike";


export default [
    Blockquote,
    BulletList,
    CodeBlock,
    Document,
    HardBreak,
    Heading,
    HorizontalRule,
    Image,
    ListItem,
    OrderedList,
    Paragraph,
    Table,
    TaskItem,
    TaskList,
    Text,

    Bold,
    Code,
    Italic,
    Link,
    Strike,
]
