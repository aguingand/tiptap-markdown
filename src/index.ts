


export { Markdown } from './Markdown';
export type { MarkdownStorage, MarkdownOptions } from './Markdown';
