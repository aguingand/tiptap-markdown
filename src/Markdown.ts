import {
    Content,
    Extension,
    ExtensionConfig,
    extensions,
    getExtensionField,
    getSchemaByResolvedExtensions, MarkConfig, NodeConfig
} from '@tiptap/core';
import { MarkdownParser } from "./MarkdownParser";
import { MarkdownClipboard } from "./extensions/markdown-clipboard/markdown-clipboard";
import markdownExtensions from "./extensions";
import { MarkdownRawHTML } from "./extensions/markdown-raw-html/markdown-raw-html";
import { MarkdownSerializer } from "./MarkdownSerializer";
import { ParseMarkdownProps } from "./types";

export interface MarkdownOptions  {
    html: boolean,
    tightLists: boolean,
    tightListClass: string,
    bulletListMarker: '-' | '*' | '+',
    horizontalRuleMarker: '-' | '_' | '*',
    linkify: boolean,
    breaks: boolean,
    transformPastedText: boolean,
    transformCopiedText: boolean,
}

export interface MarkdownStorage {
    options: MarkdownOptions,
    parser: MarkdownParser,
    serializer: MarkdownSerializer,
    getMarkdown: () => string,
    initialContent?: Content,
}

export const Markdown = Extension.create<MarkdownOptions, MarkdownStorage>({
    name: 'markdown',
    priority: 50,
    addOptions() {
        return {
            html: true,
            tightLists: true,
            tightListClass: 'tight',
            bulletListMarker: '-',
            horizontalRuleMarker: '-',
            linkify: false,
            breaks: false,
            transformPastedText: false,
            transformCopiedText: false,
        }
    },
    addCommands() {
        const commands = getExtensionField(extensions.Commands, 'addCommands')();
        return {
            setContent: (content, emitUpdate, parseOptions) => (props) => {
                return commands.setContent(
                    (props.editor.storage.markdown as MarkdownStorage).parser.parse(content),
                    emitUpdate,
                    parseOptions
                )(props);
            },
            insertContentAt: (range, content, options) => (props) => {
                return commands.insertContentAt(
                    range,
                    (props.editor.storage.markdown as MarkdownStorage).parser.parse(content),
                    options
                )(props);
            },
        }
    },
    onBeforeCreate() {
        const editor = this.editor;
        this.editor.storage.markdown = {
            options: { ...this.options },
            parser: new MarkdownParser(this.editor),
            serializer: new MarkdownSerializer(this.editor),
            getMarkdown: () => {
                return (this.editor.storage.markdown as MarkdownStorage)
                    .serializer
                    .serialize(this.editor.state.doc.content);
            },
            initialContent: this.editor.options.content,
        }
        this.editor.extensionManager.extensions = this.editor.extensionManager.extensions.map(extension => {
            let markdownExtension = markdownExtensions.find(e => e.name === extension.name);
            if(typeof markdownExtension === 'function') {
                return markdownExtension(extension).extend({
                    addStorage() {
                        return {
                            ...this.parent?.(),
                            markdown: editor.storage.markdown,
                        }
                    },
                });
            }
            if(markdownExtension) {
                return extension
                    .extend({
                        ...markdownExtension.config,
                        defaultOptions: null,
                        parseMarkdown(props) {
                            const parse = this.parent ?? getExtensionField(markdownExtension, 'parseMarkdown', this);
                            parse!(props);
                        },
                        renderMarkdown(props) {
                            const render = this.parent ?? getExtensionField(markdownExtension, 'renderMarkdown', this);
                            render!(props);
                        },
                        addStorage() {
                            return {
                                ...this.parent?.(),
                                markdown: editor.storage.markdown,
                            }
                        },
                        addOptions() {
                            return {
                                ...this.parent?.(),
                                ...markdownExtension.options,
                            }
                        },
                    } as Partial<ExtensionConfig> & Partial<NodeConfig> & Partial<MarkConfig>);
            }
            return extension;
        });
        this.editor.extensionManager.schema = getSchemaByResolvedExtensions(
            this.editor.extensionManager.extensions,
            this.editor
        );
        this.editor.schema = this.editor.extensionManager.schema;

        if(this.editor.options.content) {
            this.editor.options.content = (this.editor.storage.markdown as MarkdownStorage)
                .parser
                .parse(this.editor.options.content);
        }
    },
    onCreate() {
        this.editor.options.content = (this.editor.storage.markdown as MarkdownStorage).initialContent!;
        delete this.editor.storage.markdown.initialContent;
    },
    addStorage() {
        return {
            /// storage will be defined in onBeforeCreate() to prevent initial object overriding
        } as MarkdownStorage
    },
    addExtensions() {
        return [
            MarkdownRawHTML,
            MarkdownClipboard,
        ]
    },
});
