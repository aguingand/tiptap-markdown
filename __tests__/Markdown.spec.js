import { Editor } from "@tiptap/core";
import StarterKit from "@tiptap/starter-kit";
import { Markdown } from "../src/Markdown";
import Bold from "@tiptap/extension-bold";
import { Selection } from 'prosemirror-state';


describe('Markdown', () => {
    describe('commands', () => {
        test('setContent', () => {
            const editor = new Editor({
                extensions: [
                    Markdown,
                    StarterKit,
                ],
            });
            editor.commands.setContent('**example**');
            expect(editor.getHTML()).toBe('<p><strong>example</strong></p>');
        });
        describe('insertContentAt', () => {
            test('basic', () => {
                const editor = new Editor({
                    extensions: [
                        Markdown,
                        StarterKit,
                    ],
                });
                editor.commands.insertContentAt(Selection.atEnd(editor.state.doc), '**example**');
                expect(editor.getHTML()).toBe('<p><strong>example</strong></p>');
            });
            test('bullet list', () => {
                const editor = new Editor({
                    content: '- example1',
                    extensions: [
                        Markdown.configure({
                            tightListClass: null,
                        }),
                        StarterKit,
                    ],
                });
                editor.commands.insertContentAt(Selection.atEnd(editor.state.doc), '- example2\n  1. example3');
                expect(editor.getHTML()).toMatchSnapshot();
            });
            test('ordered list', () => {
                const editor = new Editor({
                    content: '1. example1',
                    extensions: [
                        Markdown.configure({
                            tightListClass: null,
                        }),
                        StarterKit,
                    ],
                });
                editor.commands.insertContentAt(Selection.atEnd(editor.state.doc), '  - example3');
                expect(editor.getHTML()).toMatchSnapshot();
            });
        });
    });
    test('getMarkdown', () => {
        const editor = new Editor({
            content: '<p><strong>example</strong></p>',
            extensions: [
                Markdown,
                StarterKit,
            ],
        });
        expect(editor.storage.markdown.getMarkdown()).toBe('**example**');
    });
    test('override default extension', () => {
        const editor = new Editor({
            content: '<p><strong>example</strong></p>',
            extensions: [
                Markdown,
                StarterKit.configure({ bold: false }),
                Bold.extend({
                    addStorage() {
                        return {
                            markdown: {
                                serialize: { open: '***', close: '***' },
                            }
                        }
                    }
                }),
            ],
        });
        expect(editor.storage.markdown.getMarkdown()).toBe('***example***');
    });
});
