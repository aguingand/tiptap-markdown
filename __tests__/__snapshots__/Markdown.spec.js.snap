// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown commands insertContentAt bullet list 1`] = `
<ul data-tight="true">
  <li>
    <p>
      example1
    </p>
  </li>
  <li>
    <p>
      example2
    </p>
    <ol data-tight="true">
      <li>
        <p>
          example3
        </p>
      </li>
    </ol>
  </li>
</ul>
`;

exports[`Markdown commands insertContentAt ordered list 1`] = `
<ol data-tight="true">
  <li>
    <p>
      example1
    </p>
  </li>
  <li>
    <p>
      example2
    </p>
  </li>
</ol>
<ul data-tight="true">
  <li>
    <p>
      example3
    </p>
  </li>
</ul>
`;
